### PR TITLE
chore: Update DefaultConnectionInfoCache to use the Refresher

### LINKED ISSUE #260: Make refresh algorithm like cloud-sql-jdbc-socket-factory
### Feature Description

Port the cloud-sql-jdbc-socket-factory refresh algorithm to the alloydb-java-connector.

### Sample code

_No response_

### Alternatives Considered

_No response_

### Additional Details

_No response_

### PR DESCRIPTION
**Pull request by hessjcg**
Fixes #260

### CODE REVIEW

File: alloydb-jdbc-connector/src/main/java/com/google/cloud/alloydb/Refresher.java [hunk 1]
```diff
@@ -39,7 +39,7 @@ class Refresher {
   private final Object connectionInfoGuard = new Object();
   private final AsyncRateLimiter rateLimiter;
 
-  private final RefreshCalculator refreshCalculator = new RefreshCalculator();
```
Comment (enocom) [inline alloydb-jdbc-connector/src/main/java/com/google/cloud/alloydb/Refresher.java:42]: Is this change related to the work here?

File: alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java [hunk 2]
```diff
@@ -98,348 +92,4 @@ public void testGetConnectionInfo_returnsConnectionInfo() {
     List<X509Certificate> certificateChain = connectionInfo.getCertificateChain();
     assertThat(certificateChain).hasSize(2);
   }
-
-  @Test
-  public void testGetConnectionInfo_schedulesNextOperation() {
```
Comment (ttosta-google) [inline alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java:103]: I think we should keep the tests that start with `testGetConnectionInfo_` because they validate the ConnectionInfo behavior.

I agreed that the tests starting with `testForceRefresh_` are covered by the [RefresherTest.java](https://github.com/GoogleCloudPlatform/alloydb-java-connector/blob/main/alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/RefresherTest.java).

BTW could you please rename the tests in `RefresherTest` ? Some of them starts with `testCloudSql`

File: alloydb-jdbc-connector/src/main/java/com/google/cloud/alloydb/Refresher.java [hunk 3]
```diff
@@ -55,21 +55,31 @@ class Refresher {
   @GuardedBy("connectionInfoGuard")
   private Throwable currentRefreshFailure;
 
+  Refresher(
+      String name,
+      ListeningScheduledExecutorService executor,
+      Supplier<ListenableFuture<ConnectionInfo>> refreshOperation,
+      AsyncRateLimiter rateLimiter) {
+    this(name, executor, new RefreshCalculator(), refreshOperation, rateLimiter);
+  }
```
Comment (enocom) [inline alloydb-jdbc-connector/src/main/java/com/google/cloud/alloydb/Refresher.java:64]: This looks like an unrelated change. Otherwise the PR LGTM.

**Review by enocom — CHANGES_REQUESTED** [review]

File: alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java (hunk 2 above)
Comment (hessjcg) [reply alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java:103]: I updated `RefresherTest` to remove "CloudSql" from the testcase names. 
I restored some of the test cases to `ConnectionInfoCacheTest`. I'll leave inline comments about why some testcases were modified or removed.

File: alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java [hunk 4]
```diff
@@ -43,26 +43,32 @@ public class ConnectionInfoCacheTest {
...
 
   @Before
   public void setUp() throws CertificateException, IOException, OperatorCreationException {
     instanceName =
         InstanceName.parse(
             "projects/<PROJECT>/locations/<REGION>/clusters/<CLUSTER>/instances/<INSTANCE>");
     keyPair = RsaKeyPairGenerator.generateKeyPair();
-    spyRateLimiter = new SpyRateLimiter();
     testCertificates = new TestCertificates();
+    ScheduledThreadPoolExecutor exec = new ScheduledThreadPoolExecutor(4);
```
Comment (hessjcg) [inline alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java:59]: The Refresher tests only work with a real ScheduledThreadPoolExecutor. We can't use DeterministicExecutor.

File: alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java (hunk 2 above)
Comment (hessjcg) [reply alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java:103]: testGetConnectionInfo_schedulesNextOperation was removed because it relied on DeterministicExecutor. It is covered by `RefresherTest.testRefreshesExpiredData()`

File: alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java [hunk 5]
```diff
@@ -193,14 +141,13 @@ public Object getTransportCode() {
             connectionInfoRepo,
             instanceName,
             keyPair,
-            new RefreshCalculator(),
-            spyRateLimiter);
+            TEST_TIMEOUT_MS);
 
-    executor.runNextPendingCommand(); // Simulate completion of background thread.
```
Comment (hessjcg) [inline alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java:199]: The Refresher will not throw an exception when a test fails. Instead it will automatically retry, complying with the rate limiter. This is covered by `RefresherTest.testRefreshRetriesOnAfterFailedAttempts()`

File: alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java [hunk 6]
```diff
@@ -215,7 +162,7 @@ public Object getTransportCode() {
 
   @Test
   public void testGetConnectionInfo_scheduledNextOperationImmediately_onCertificateException() {
-    DeterministicScheduler executor = new DeterministicScheduler();
```
Comment (hessjcg) [inline alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java:218]: The Refresher will not throw an exception when a test fails. Instead it will automatically retry, complying with the rate limiter. This is covered by `RefresherTest.testRefreshRetriesOnAfterFailedAttempts()`

File: alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java [hunk 7]
```diff
@@ -237,209 +184,22 @@ public void testGetConnectionInfo_scheduledNextOperationImmediately_onCertificat
...
-            connectionInfo
-                .getClientCertificate()
-                .getNotAfter()
-                .toInstant()
-                .truncatedTo(ChronoUnit.SECONDS))
-        .isEqualTo(ONE_HOUR_FROM_NOW.truncatedTo(ChronoUnit.SECONDS));
-  }
-
-  @Test
-  public void testGetConnectionInfo_isRateLimited() {
```
Comment (hessjcg) [inline alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java:265]: The rate limiter is internal to the Refresher and can't be tested here. I added `RefresherTest.testRateLimiterInUse()` to test this.

File: alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java [hunk 8]
```diff
@@ -237,209 +184,22 @@ public void testGetConnectionInfo_scheduledNextOperationImmediately_onCertificat
...
-
-    executor.runNextPendingCommand(); // Simulate completion of background thread.
-    connectionInfoCache.getConnectionInfo();
+            TEST_TIMEOUT_MS);
 
-    assertThat(spyRateLimiter.wasRateLimited.get()).isTrue();
-  }
-
-  @Test
-  public void testForceRefresh_schedulesNextRefreshImmediately() throws InterruptedException {
```
Comment (hessjcg) [inline alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java:296]: This is tested by `RefresherTest.testForcesRefresh()`

File: alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java [hunk 9]
```diff
@@ -237,209 +184,22 @@ public void testGetConnectionInfo_scheduledNextOperationImmediately_onCertificat
...
-                .getClientCertificate()
-                .getNotAfter()
-                .toInstant()
-                .truncatedTo(ChronoUnit.SECONDS))
-        .isEqualTo(TWO_HOURS_FROM_NOW.truncatedTo(ChronoUnit.SECONDS));
-  }
 
-  @Test
-  public void testForceRefresh_refreshCalledOnlyOnceDuringMultipleCalls()
-      throws InterruptedException {
```
Comment (hessjcg) [inline alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java:368]: This is tested by `RefresherTest.testThatForceRefreshBalksWhenAScheduledRefreshIsInProgress()` and `RefresherTest.testThatForceRefreshBalksWhenAForceRefreshIsInProgress()`

File: alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java [hunk 10]
```diff
@@ -193,14 +141,13 @@ public Object getTransportCode() {
...
-            spyRateLimiter);
+            TEST_TIMEOUT_MS);
 
-    executor.runNextPendingCommand(); // Simulate completion of background thread.
-    assertThrows(RuntimeException.class, connectionInfoCache::getConnectionInfo);
-
-    executor.tick(1, TimeUnit.SECONDS); // Advance time just a little
-    executor.runUntilIdle(); // Simulate completion of background thread.
+    try {
+      Thread.sleep(1000);
```
Comment (ttosta-google) [inline alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java:147]: why is it necessary?

Comment (hessjcg) [reply alloydb-jdbc-connector/src/test/java/com/google/cloud/alloydb/ConnectionInfoCacheTest.java:147]: We need to allow the refresh process to run in the background and finish before we can test the assertions. RefresherTest has a more elegant way to handle this. But it's not available to us in the ConnectionInfoCacheTest.

**Review by ttosta-google — APPROVED** [review]
LGTM
